File: Java/707.java
__________________________________________________________________________________________________
sample 47 ms submission
class MyLinkedList {
    private int start;
    private int size;
    private int capacity;
    private int[] storage;

    private void adjustCapacity(int capacity) {
        int[] storage = new int[capacity];
        System.arraycopy(this.storage, start, storage, 0, this.size - start);
        System.arraycopy(this.storage, 0, storage, this.size - start, start);
        start = 0;

        this.storage = storage;
        this.capacity = capacity;
    }

    /**
     * Initialize your data structure here.
     */
    public MyLinkedList() {
        start = 0;
        capacity = 1;
        storage = new int[capacity];
    }

    /**
     * Get the value of the index-th node in the linked list. If the index is invalid, return -1.
     */
    public int get(int index) {
        if (index < 0 || index >= size) return -1;
        return storage[(start + index) % capacity];
    }

    /**
     * Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list.
     */
    public void addAtHead(int val) {
        addAtIndex(0, val);
    }

    /**
     * Append a node of value val to the last element of the linked list.
     */
    public void addAtTail(int val) {
        addAtIndex(size, val);
    }

    /**
     * Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted.
     */
    public void addAtIndex(int index, int val) {
        if (index > size) return;
        if (index < 0) index = 0;

        if (index == 0) {
            start = (start + capacity - 1) % capacity;
            storage[start] = val;
        } else if (index == size) {
            storage[(start + size) % capacity] = val;
        } else if (index >= size / 2) {
            if (start + size < capacity) {
                System.arraycopy(storage, start + index, storage, start + index + 1, size - index);
            } else if (start + index >= capacity) {
                System.arraycopy(storage, (start + index) % capacity, storage, (start + index + 1) % capacity, size - index);
            } else {
                int t = storage[capacity - 1];
                System.arraycopy(storage, start + index, storage, start + index + 1, capacity - start - index - 1);
                System.arraycopy(storage, 0, storage, 1, size - capacity + start + index + 1);
                storage[0] = t;
            }
            storage[(start + index) % capacity] = val;
        } else {
            if (start == 0) {
                storage[capacity - 1] = storage[0];
                System.arraycopy(storage, 1, storage, 0, index - 1);
            } else if (start + index <= capacity) {
                System.arraycopy(storage, start, storage, start - 1, index);
            } else {
                System.arraycopy(storage, start, storage, start - 1, capacity - start);
                storage[capacity - 1] = storage[0];
                System.arraycopy(storage, 1, storage, 0, index - capacity + start);
            }

            storage[(start + index - 1) % capacity] = val;
            start = (start + capacity - 1) % capacity;
        }

        size++;

        if (size == capacity) adjustCapacity(capacity * 2);
    }

    /**
     * Delete the index-th node in the linked list, if the index is valid.
     */
    public void deleteAtIndex(int index) {
        if (index < 0 || index >= size) return;

        if (index >= size / 2) {
            if (start + size <= capacity) {
                System.arraycopy(storage, start + index + 1, storage, start + index, size - index - 1);
            } else if (start + index >= capacity) {
                System.arraycopy(storage, start + index + 1 - capacity, storage, start + index - capacity, size - index - 1);
            } else {
                System.arraycopy(storage, start + index + 1, storage, start + index, capacity - start - index - 1);
                storage[capacity - 1] = storage[0];
                System.arraycopy(storage, 1, storage, 0, start + size - capacity - 1);
            }
            storage[(start + size + capacity - 1) % capacity] = 0;
        } else {
            if (start + index < capacity) {
                System.arraycopy(storage, start, storage, start + 1, index);
            } else {
                int t = storage[capacity - 1];
                System.arraycopy(storage, start, storage, start + 1, capacity - start - 1);
                System.arraycopy(storage, 0, storage, 1, index - capacity + start);
                storage[0] = t;
            }
            storage[start] = 0;
            start = (start + 1) % capacity;
        }

        size--;
        if (size == capacity / 4) adjustCapacity((capacity / 2));
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.get(index);
 * obj.addAtHead(val);
 * obj.addAtTail(val);
 * obj.addAtIndex(index,val);
 * obj.deleteAtIndex(index);
 */
__________________________________________________________________________________________________
sample 42792 kb submission
class MyLinkedList {

   List<Integer> list;
    int len;
    /** Initialize your data structure here. */
    public MyLinkedList() {
        list = new ArrayList<>();
        len = 0;
    }
    
    /** Get the value of the index-th node in the linked list. If the index is invalid, return -1. */
    public int get(int index) {
        if(index < 0 || index >= len || len == 0) return -1;
        return list.get(index);
    }
    
    /** Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list. */
    public void addAtHead(int val) {
        list.add(0,val);
        len++;
    }
    
    /** Append a node of value val to the last element of the linked list. */
    public void addAtTail(int val) {
        list.add(val);
        len++;
    }
    
    /** Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted. */
    public void addAtIndex(int index, int val) {
        if(index > len) return;
        if(index < 0){
            list.add(0,val);
        }else if(index == len){
            list.add(val);  
        } else{
            list.add(index,val);
        }
        len++;
    }
    
    /** Delete the index-th node in the linked list, if the index is valid. */
    public void deleteAtIndex(int index) {
        if(index < 0 || index >= len || len == 0) return;
        list.remove(index);
        len--;
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.get(index);
 * obj.addAtHead(val);
 * obj.addAtTail(val);
 * obj.addAtIndex(index,val);
 * obj.deleteAtIndex(index);
 */
__________________________________________________________________________________________________

File: Python3/707.py
__________________________________________________________________________________________________
sample 84 ms submission
class MyLinkedList:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.stack = []
        

    def get(self, index: int) -> int:
        """
        Get the value of the index-th node in the linked list. If the index is invalid, return -1.
        """
        if index >= len(self.stack) or index < 0:
            return -1
        else:
            return self.stack[index]
        

    def addAtHead(self, val: int) -> None:
        """
        Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list.
        """
        self.stack.insert(0, val)
        

    def addAtTail(self, val: int) -> None:
        """
        Append a node of value val to the last element of the linked list.
        """
        self.stack.append(val)
        

    def addAtIndex(self, index: int, val: int) -> None:
        """
        Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted.
        """
        if len(self.stack) < index:
            return -1
        
        self.stack.insert(index, val)
        

    def deleteAtIndex(self, index: int) -> None:
        """
        Delete the index-th node in the linked list, if the index is valid.
        """
        if index >= len(self.stack) or index < 0:
            return -1
        else:
            self.stack.pop(index)
        


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)
__________________________________________________________________________________________________
sample 13368 kb submission
class Node(object):
    def __init__(self):
        self.val = None
        self.next = None


class MyLinkedList:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.count = 0
        self.root = None
        

    def get(self, index: int) -> int:
        """
        Get the value of the index-th node in the linked list. If the index is invalid, return -1.
        """
        if index >= self.count or index < 0:
            return -1
        
        head = self.root
        for i in range(index):
            head = head.next
        # print(index, self.count, self.root)
        return head.val

    def addAtHead(self, val: int) -> None:
        """
        Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list.
        """
        node = Node()
        node.val = val
        node.next = self.root
        self.root = node
        self.count += 1

    def addAtTail(self, val: int) -> None:
        """
        Append a node of value val to the last element of the linked list.
        """
        node = Node()
        node.val = val
        
        if self.count == 0:
            self.root = node
        else:
            head = self.root
            for _ in range(self.count-1):
                head = head.next
            
            head.next = node

        self.count += 1
        

    def addAtIndex(self, index: int, val: int) -> None:
        """
        Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted.
        """
        
        if index > self.count:
            return
        
        node = Node()
        node.val = val        
        
        if self.count == 0:
            node.next = self.root
            self.root = node
            self.count += 1
            return
        
        if self.count == index:
            self.addAtTail(val)
        else:

            if index == 0:
                self.addAtHead(val)
            else:         
                head = self.root
                for _ in range(index-1):
                    head = head.next

                node.next = head.next
                head.next = node
#             node_before = self.getNode(index-1)
#             node_after = self.getNode(index)

#             new_node = Node()
#             node.val = val

#             node_before.next = new_node
#             new_node.next = node_after
                
                self.count += 1
                
            
    def deleteAtIndex(self, index: int) -> None:
        """
        Delete the index-th node in the linked list, if the index is valid.
        """
        
        if index < 0 or index >= self.count:
            return
        
        node_to_delete = self.root
        for _ in range(index):
            node_to_delete = node_to_delete.next
        
        del node_to_delete
        
        if index == 0:
            new_head = self.getNode(1)
            self.root = new_head
        else:
            node_before = self.getNode(index-1)
            node_after = self.getNode(index+1)
            node_before.next = node_after
        self.count -= 1
        
    def getNode(self, index: int) -> ListNode:
        cur = self.root
        for _ in range(index):
            cur = cur.next
        return cur
    
#     def get(self, index: 'int') -> 'int':
#         if index < 0 or index >= self.count:
#             return -1
        
#         return self.getNode(index).val


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)
__________________________________________________________________________________________________
